File: app/src/main/java/com/maafw/naruto/bridge/NativeBridgeLib.java
package com.maafw.naruto.bridge;

import android.annotation.SuppressLint;
import android.graphics.Bitmap;
import android.view.Surface;

/**
 * libbridge.so 的 JNI 桥（类名必须匹配原生 JNI 符号）。
 *
 * ⚠ 重要：libbridge.so 不可重编译（原版 md5 6807aea6，重编译会导致滑动卡死/启动崩溃），
 * 其 JNI 导出符号基于本类名（com_maafw_naruto_bridge_NativeBridgeLib），
 * 因此**本类名不可改**。业务代码请统一使用 {@link BridgeNativeLib}（封装本类）。
 */
@SuppressLint("PrivateApi,DiscouragedPrivateApi")
public final class NativeBridgeLib {

    private NativeBridgeLib() {
    }

    /** 探活 */
    public static native String ping();

    /** 建立原生捕获器，返回 Surface 投给虚拟屏 */
    public static native Surface setupNativeCapturer(int width, int height);

    /** 释放捕获器 */
    public static native void releaseNativeCapturer();

    /** 设置预览 Surface */
    public static native void setPreviewSurface(Object surface);

    /** 取当前帧 Bitmap（测试/截图用） */
    public static native Bitmap getFrameBufferBitmap();

    /** 帧计数（等首帧/黑屏判断） */
    public static native long getFrameCount();

    /** 游戏 FPS（原生统计） */
    public static native double getFps();

    /** 脚本识别频率（原生统计） */
    public static native double getScriptFps();

    /** 切换进程 UID（root 引擎发广播前降权/发完提权） */
    public static native int setResUid(int ruid, int euid, int suid);
}

File: app/src/main/java/com/maafw/naruto/bridge/NativeBridgeLib.kt
package com.maafw.naruto.bridge

import android.graphics.Bitmap
import android.view.Surface

/**
 * 原生桥 Java 层入口喵～
 * 对应 libbridge.so 中 JNI 注册的 NativeBridgeLib 类。
 */
object NativeBridgeLib {

    @JvmField
    var LOADED: Boolean = false

    init {
        LOADED = try {
            System.loadLibrary("bridge")
            true
        } catch (e: Throwable) {
            false
        }
    }

    @JvmStatic
    external fun ping(): String

    @JvmStatic
    external fun setupNativeCapturer(width: Int, height: Int): Surface?

    @JvmStatic
    external fun releaseNativeCapturer()

    @JvmStatic
    external fun setPreviewSurface(surface: Any?)

    @JvmStatic
    external fun getFrameBufferBitmap(): Bitmap

    @JvmStatic
    external fun getFrameCount(): Long

    /**
     * 切换进程 UID（root 引擎发广播前临时降权到 App uid，发完提权回 root）喵。
     * @return setresuid 返回值（0=成功）
     */
    @JvmStatic
    external fun setResUid(ruid: Int, euid: Int, suid: Int): Int
}